File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/compendium.ts
/** Monster and item compendium: create-and-fetch reference data backed by SQLite directly (no in-memory cache). */

import { getDb } from '../db.ts';
import type { ApiResult, JsonValue } from '../types.ts';
import { isValidIntInRange } from '../validation.ts';

const SLUG_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

interface MonsterRow {
  slug: string;
  name: string;
  cr: string;
  armor_class: number;
  hit_points: number;
  tags_json: string;
}

function monsterRowToBody(row: MonsterRow): JsonValue {
  return {
    slug: row.slug,
    name: row.name,
    cr: row.cr,
    armor_class: row.armor_class,
    hit_points: row.hit_points,
    tags: JSON.parse(row.tags_json),
  };
}

export function createMonster(body: JsonValue): ApiResult {
  const slug = body.slug;
  if (typeof slug !== 'string' || !SLUG_RE.test(slug)) {
    return { status: 400, body: { error: 'slug must be a lowercase kebab-case string' } };
  }

  const name = body.name;
  if (typeof name !== 'string' || name.length === 0) {
    return { status: 400, body: { error: 'name must be a non-empty string' } };
  }

  const cr = body.cr;
  if (typeof cr !== 'string' || cr.length === 0) {
    return { status: 400, body: { error: 'cr must be a non-empty string' } };
  }

  const armorClass = body.armor_class;
  if (!isValidIntInRange(armorClass, 0, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'armor_class must be a non-negative integer' } };
  }

  const hitPoints = body.hit_points;
  if (!isValidIntInRange(hitPoints, 0, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'hit_points must be a non-negative integer' } };
  }

  const tags = body.tags ?? [];
  if (!Array.isArray(tags) || !tags.every((t) => typeof t === 'string')) {
    return { status: 400, body: { error: 'tags must be an array of strings' } };
  }

  const db = getDb();
  const existing = db.prepare('SELECT slug FROM monsters WHERE slug = ?').get(slug);
  if (existing) {
    return { status: 409, body: { error: 'monster slug already exists' } };
  }

  db.prepare(
    'INSERT INTO monsters (slug, name, cr, armor_class, hit_points, tags_json) VALUES (?, ?, ?, ?, ?, ?)',
  ).run(slug, name, cr, armorClass, hitPoints, JSON.stringify(tags));

  return {
    status: 201,
    body: { slug, name, cr, armor_class: armorClass, hit_points: hitPoints },
  };
}

export function getMonster(slug: string): ApiResult {
  const db = getDb();
  const row = db.prepare('SELECT slug, name, cr, armor_class, hit_points, tags_json FROM monsters WHERE slug = ?').get(
    slug,
  ) as MonsterRow | undefined;
  if (!row) {
    return { status: 404, body: { error: 'unknown monster slug' } };
  }
  return { status: 200, body: monsterRowToBody(row) };
}

interface ItemRow {
  slug: string;
  name: string;
  type: string;
  rarity: string;
  cost_gp: number;
}

function itemRowToBody(row: ItemRow): JsonValue {
  return {
    slug: row.slug,
    name: row.name,
    type: row.type,
    rarity: row.rarity,
    cost_gp: row.cost_gp,
  };
}

export function createItem(body: JsonValue): ApiResult {
  const slug = body.slug;
  if (typeof slug !== 'string' || !SLUG_RE.test(slug)) {
    return { status: 400, body: { error: 'slug must be a lowercase kebab-case string' } };
  }

  const name = body.name;
  if (typeof name !== 'string' || name.length === 0) {
    return { status: 400, body: { error: 'name must be a non-empty string' } };
  }

  const type = body.type;
  if (typeof type !== 'string' || type.length === 0) {
    return { status: 400, body: { error: 'type must be a non-empty string' } };
  }

  const rarity = body.rarity;
  if (typeof rarity !== 'string' || rarity.length === 0) {
    return { status: 400, body: { error: 'rarity must be a non-empty string' } };
  }

  const costGp = body.cost_gp;
  if (typeof costGp !== 'number' || !Number.isFinite(costGp) || costGp < 0) {
    return { status: 400, body: { error: 'cost_gp must be a non-negative number' } };
  }

  const db = getDb();
  const existing = db.prepare('SELECT slug FROM items WHERE slug = ?').get(slug);
  if (existing) {
    return { status: 409, body: { error: 'item slug already exists' } };
  }

  db.prepare('INSERT INTO items (slug, name, type, rarity, cost_gp) VALUES (?, ?, ?, ?, ?)').run(
    slug,
    name,
    type,
    rarity,
    costGp,
  );

  return {
    status: 201,
    body: { slug, name, type, rarity, cost_gp: costGp },
  };
}

export function getItem(slug: string): ApiResult {
  const db = getDb();
  const row = db.prepare('SELECT slug, name, type, rarity, cost_gp FROM items WHERE slug = ?').get(
    slug,
  ) as ItemRow | undefined;
  if (!row) {
    return { status: 404, body: { error: 'unknown item slug' } };
  }
  return { status: 200, body: itemRowToBody(row) };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/compendium.rb
# Reference data catalogs: monsters and items. Create-then-read only, keyed
# by a caller-supplied slug; no update/delete endpoints exist.

post '/v1/compendium/monsters' do
  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  slug = body['slug']
  name = body['name']
  cr = body['cr']
  armor_class = body['armor_class']
  hit_points = body['hit_points']
  tags = body['tags']
  tags = [] if tags.nil?

  halt 400, { error: 'invalid slug' }.to_json unless valid_slug?(slug)
  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?
  halt 400, { error: 'invalid cr' }.to_json unless cr.is_a?(String) && !cr.empty?
  halt 400, { error: 'invalid armor_class' }.to_json unless integerish(armor_class)
  halt 400, { error: 'invalid hit_points' }.to_json unless integerish(hit_points)
  halt 400, { error: 'invalid tags' }.to_json unless tags.is_a?(Array) && tags.all? { |t| t.is_a?(String) }
  halt 409, { error: 'slug already exists' }.to_json if db.execute('SELECT 1 FROM monsters WHERE slug = ?', [slug]).first

  armor_class = armor_class.to_i
  hit_points = hit_points.to_i

  db.execute(
    'INSERT INTO monsters (slug, name, cr, armor_class, hit_points, tags_json) VALUES (?, ?, ?, ?, ?, ?)',
    [slug, name, cr, armor_class, hit_points, tags.to_json]
  )

  status 201
  {
    slug: slug,
    name: name,
    cr: cr,
    armor_class: armor_class,
    hit_points: hit_points
  }.to_json
end

get '/v1/compendium/monsters/:slug' do
  row = db.execute('SELECT * FROM monsters WHERE slug = ?', [params[:slug]]).first
  halt 404, { error: 'unknown monster' }.to_json unless row

  {
    slug: row['slug'],
    name: row['name'],
    cr: row['cr'],
    armor_class: row['armor_class'],
    hit_points: row['hit_points'],
    tags: JSON.parse(row['tags_json'])
  }.to_json
end

post '/v1/compendium/items' do
  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  slug = body['slug']
  name = body['name']
  type = body['type']
  rarity = body['rarity']
  cost_gp = body['cost_gp']

  halt 400, { error: 'invalid slug' }.to_json unless valid_slug?(slug)
  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?
  halt 400, { error: 'invalid type' }.to_json unless type.is_a?(String) && !type.empty?
  halt 400, { error: 'invalid rarity' }.to_json unless rarity.is_a?(String) && !rarity.empty?
  halt 400, { error: 'invalid cost_gp' }.to_json unless integerish(cost_gp)
  halt 409, { error: 'slug already exists' }.to_json if db.execute('SELECT 1 FROM items WHERE slug = ?', [slug]).first

  cost_gp = cost_gp.to_i

  db.execute(
    'INSERT INTO items (slug, name, type, rarity, cost_gp) VALUES (?, ?, ?, ?, ?)',
    [slug, name, type, rarity, cost_gp]
  )

  status 201
  {
    slug: slug,
    name: name,
    type: type,
    rarity: rarity,
    cost_gp: cost_gp
  }.to_json
end

get '/v1/compendium/items/:slug' do
  row = db.execute('SELECT * FROM items WHERE slug = ?', [params[:slug]]).first
  halt 404, { error: 'unknown item' }.to_json unless row

  {
    slug: row['slug'],
    name: row['name'],
    type: row['type'],
    rarity: row['rarity'],
    cost_gp: row['cost_gp']
  }.to_json
end
